File: 258.cpp
#include <iostream>
#include <cstring>
int64_t matrix[2000][2000];
int64_t result[2000][2000];
int64_t mod = 20092010;
int64_t initial[2000];

void matrix_multiply(int64_t matrix1[2000][2000], int64_t matrix2[2000][2000]) {

    for(int i = 0 ; i < 2000; ++i) {
        for(int j = 0; j < 2000; ++j) {
            result[i][j] = 0;
            for(int k = 0; k < 2000; ++k) {
                result[i][j] += matrix2[i][k] * matrix2[k][j];
                result[i][j] %= mod;
            }
        }
    }

    memcpy(matrix1, result, sizeof(result));

}

void matrix_expo(int64_t matrix[2000][2000], int64_t n) {
    for(int i = 0; i < 2000; ++i) {
        for(int j = 0; j < 2000; ++j) {
            result[i][j] = matrix[i][j];
        }
    }

    while(n > 0) {
        std::cout << n << std::endl;
        if(n % 2 == 1) {
            matrix_multiply(result, matrix);
        }
        matrix_multiply(matrix, matrix);
        n /= 2;
    }
}

int main() {
    matrix[2000-1][0] = 1;
    matrix[2000-2][0] = 1;

    for(int row = 0; row + 1 < 2000; ++row) {
       matrix[row][row+1] = 1;
    }

    matrix_expo(matrix, 20000);


    int64_t ans = 0;
    for(int i = 0; i < 2000; ++i) {
        ans += matrix[0][i];
        ans %= mod;
    }

    std::cout << ans << std::endl;
}

File: 258.py
import numpy as np
from scipy import sparse

# Define constants
SIZE = 2000
MOD = 20092010

def create_sparse_matrix():
    """Create the sparse matrix for the recurrence relation."""
    # Create the sparse matrix using coordinate format (COO)
    row_indices = []
    col_indices = []
    values = []
    
    # Add the two ones in the first column
    row_indices.extend([SIZE-1, SIZE-2])
    col_indices.extend([0, 0])
    values.extend([1, 1])
    
    # Add the shifted diagonal ones
    for i in range(SIZE-1):
        row_indices.append(i)
        col_indices.append(i+1)
        values.append(1)
    
    # Create the sparse matrix
    return sparse.csr_matrix((values, (row_indices, col_indices)), shape=(SIZE, SIZE))

def sparse_matrix_multiply_mod(A, B, mod):
    """Multiply two sparse matrices with modular arithmetic."""
    C = A @ B
    # Apply modulo to non-zero elements
    C.data %= mod
    return C

def sparse_matrix_expo(matrix, n):
    """Compute matrix^n using binary exponentiation with sparse matrices."""
    result = matrix.copy()
    power = matrix.copy()
    n -= 1  # Adjust because we start with result = matrix
    
    while n > 0:
        print(f"Remaining power: {n}")
        if n % 2 == 1:
            result = sparse_matrix_multiply_mod(result, power, MOD)
        power = sparse_matrix_multiply_mod(power, power, MOD)
        n //= 2
    
    return result

def main():
    # Create the sparse matrix
    matrix = create_sparse_matrix()
    
    # Compute matrix^20000
    result = sparse_matrix_expo(matrix, 10**7)
    
    # Sum the first row elements and apply modulo
    first_row = result[0, :].toarray()[0]
    ans = sum(first_row) % MOD
    
    print(f"Final answer: {ans}")

if __name__ == "__main__":
    main()